pkg/config/store/store2.go

```diff
@@ -31,47 +34,154 @@ type Key struct {
+
+	// Watch creates a channel to receive the events.
+	Watch(ctx context.Context) (<-chan BackendEvent, error)
+
+	// Get returns a resource's spec to the key.
+	Get(key Key) (map[string]interface{}, error)
+
+	// List returns the whole mapping from key to resource specs in the store.
+	List() map[Key]map[string]interface{}
+}
+
 // Store2 defines the access to the storage for mixer.
 // TODO: rename to Store.
 type Store2 interface {
-	// SetValidator sets the validator for the store.
-	SetValidator(v Validator)
-
-	// RegisterKind registers a kind to be structured as the given proto.Message.
-	// This will be used for structural validations.
-	RegisterKind(kind string, spec proto.Message)
-
-	// Init initializes the connection with the storage backend. Any further registrations
-	// of kind through RegisterKind() will be ignored after Init(). The connection will
-	// be closed after ctx is done.
-	Init(ctx context.Context) error
+	Init(ctx context.Context, kinds map[string]proto.Message) error
 
-	// Watch creates a channel to receive the events on the given kinds.
-	Watch(ctx context.Context, kinds []string) (<-chan Event, error)
+	// Watch creates a channel to receive the events.
+	Watch(ctx context.Context) (<-chan Event, error)
 
 	// Get returns a resource's spec to the key.
 	Get(key Key, spec proto.Message) error
 
 	// List returns the whole mapping from key to resource specs in the store.
 	List() map[Key]proto.Message
+}
+
+// store2 is the implementation of Store2 interface.
+type store2 struct {
+	kinds   map[string]proto.Message
+	backend Store2Backend
+}
+
+// SetValidator sets the validator for the store.
+func (s *store2) SetValidator(v Validator) {
+	// TODO: implement this
+}
+
+// Init initializes the connection with the storage backend. This uses "kinds"
+// for the mapping from the kind's name and its structure in protobuf.
+// The connection will be closed after ctx is done.
+func (s *store2) Init(ctx context.Context, kinds map[string]proto.Message) error {
+	kindNames := make([]string, 0, len(kinds))
+	for k := range kinds {
+		kindNames = append(kindNames, k)
+	}
+	if err := s.backend.Init(ctx, kindNames); err != nil {
+		return err
+	}
+	s.kinds = kinds
+	return nil
+}
+
+// Watch creates a channel to receive the events.
+func (s *store2) Watch(ctx context.Context) (<-chan Event, error) {
+	ch, err := s.backend.Watch(ctx)
+	if err != nil {
+		return nil, err
+	}
+	q := newQueue(ctx, ch, s.kinds)
+	return q.chout, nil
+}
+
+// Get returns a resource's spec to the key.
+func (s *store2) Get(key Key, spec proto.Message) error {
+	obj, err := s.backend.Get(key)
+	if err != nil {
+		return err
+	}
+
+	return convert(obj, spec)
+}
+
+// List returns the whole mapping from key to resource specs in the store.
+func (s *store2) List() map[Key]proto.Message {
+	data := s.backend.List()
+	result := make(map[Key]proto.Message, len(data))
+	for k, spec := range data {
+		pbSpec, err := convertWithKind(spec, k.Kind, s.kinds)
```

Comment:
The main point here is to create a clone of the message from `s.kinds`, and then convert `spec` into the clone (otherwise `s.kinds` will be affected). Also, the same logic is used in `queue.go`.

Maybe this structure wasn't clear. removed `convertWithKind`, create `cloneMessage` utility instead.